paddle/framework/data_transform.h

```diff
@@ -45,6 +48,42 @@ struct KernelTypePairHash {
   }
 };
 
+template <typename InType, typename OutType>
+struct CastDataTypeFunctor {
+  HOSTDEVICE OutType operator()(InType in) const {
+    return static_cast<OutType>(in);
+  }
+};
+
+template <typename InType>
+struct CastDataType {
+  CastDataType(const framework::Tensor& in, framework::Tensor* out,
+               const platform::DeviceContext* ctx)
+      : in_(in), out_(out), ctx_(ctx) {}
+  const framework::Tensor in_;
+  framework::Tensor* out_;
+  const platform::DeviceContext* ctx_;
+
+  template <typename OutType>
+  void operator()() {
+    auto place = ctx_->GetPlace();
+
+    auto* in_begin = in_.data<InType>();
+    auto numel = in_.numel();
+    auto* in_end = in_begin + numel;
+    auto* out_begin = out_->mutable_data<OutType>(place);
+    if (platform::is_cpu_place(place)) {
+      platform::Transform<platform::CPUDeviceContext> trans;
+      auto* context = static_cast<const platform::CPUDeviceContext*>(ctx_);
+      trans(*context, in_begin, in_end, out_begin,
+            CastDataTypeFunctor<InType, OutType>());
+    } else {
+      // TODO(dzhwinter): enhance CopyFrom CPU<->GPU with different data type?
+      PADDLE_THROW("Unsupport CPU <-> GPU!");
```

Comment:
`OpKernelType` is made up of four members(type, layout, place, library), except for library, the other three members are related to Data. According to the principle of function: a single task, I suggest that `CastDataType` only handles the data type conversion on the same device.

Reply:
Yes, we have re-designed the registrars.

Reply:
In the very beginning, we didn't figure out a method to do the multiple transforms.

For example, if we need to transform 0000 -> 1111, should we done in 4 step or 1 step?

>  *  For DataType, only test the FP32(float), FP64(double).
 *  e.g. 0000 -> FP32, CPUPlace, kNHWC, kPlain
 *       1111 -> FP64, GPUPlace, kNCHW, kMKLDNN
So this PR looks ugly when deal with CPU <-> GPU, and transform datatype the same time.

Now we have made the decision, we will do this in 4 step. ONE functor will only do its own duty, then we will only have four kinds of transform functor. We will achieve that in the final state.

Reply:
For `LibraryType`, `Variable` does not need a transform functor, right?

Reply:
For LibraryType, Variable does not need a transform functor, right?

Reply:
Yes, only 3 kinds of transform functor if you take it seriously.